File: ml_tools/ui/service/services.py
import threading

from ..protocol import Message, Protocol

from .control import Control
from .video import Video


class Services:
  def __init__(self, protocol: Protocol):
    self.protocol = protocol

    self.control = Control(protocol)
    self.video = Video(protocol)

    self.services = [self.control, self.video]

  def start(self):
    threading.Thread(target=self.start_reading_loop).start()

  def start_reading_loop(self):
    while True:
      message = self.protocol.Read()
      for service in self.services:
        (type_min, type_max) = service.message_type_range
        if message.Type >= type_min and message.Type <= type_max:
          service.on_message(message)
          break

File: ml_tools/ui/server/service/services.go
package service

import (
	"io"
	"ml_tools/server/protocol"
)

type Service interface {
	// OnMessage must be synchronized, so it could be called from
	// multiple go-rutines.
	OnMessage(message protocol.Message)
	GetMessageTypeRange() (min, max int)
}

type Services struct {
	rw      protocol.ReadWriter
	Control *Control
	Video   *Video

	services []Service
}

func NewServies(rw protocol.ReadWriter) *Services {
	s := &Services{}
	s.rw = rw

	s.Control = &Control{rw}
	s.Video = newVideo(rw)

	s.services = []Service{s.Control, s.Video}
	return s
}

func (s *Services) Start() {
	for i := 0; i < 3; i++ {
		go s.startReadingLoop()
	}
}

func (s *Services) startReadingLoop() {
	for {
		message, err := s.rw.Read()
		if err != nil {
			if err == io.EOF {
				return
			}
			panic(err)
		}

		for _, service := range s.services {
			min, max := service.GetMessageTypeRange()
			if int(message.Type) >= min && int(message.Type) <= max {
				service.OnMessage(message)
				break
			}
		}
	}
}
